cmd/boulder-sa/main.go

```diff
@@ -77,16 +79,41 @@ func main() {
 	dbMap, err := sa.NewDbMap(dbURL, saDbSettings)
 	cmd.FailOnError(err, "Couldn't connect to SA database")
 
+	dbAddr, dbUser, err := saConf.DB.DSNAddressAndUser()
+	cmd.FailOnError(err, "Could not determine address or user of DB DSN")
+
 	// Collect and periodically report DB metrics using the DBMap and prometheus scope.
-	sa.InitDBMetrics(dbMap, scope, saDbSettings)
+	sa.InitDBMetrics(dbMap, scope, saDbSettings, dbAddr, dbUser)
+
+	var dbReadOnlyMap *db.WrappedMap
+
+	dbReadOnlyURL, err := saConf.ReadOnlyDB.URL()
+	cmd.FailOnError(err, "Couldn't load read-only DB URL")
+	if dbReadOnlyURL == "" {
+		dbReadOnlyMap = dbMap
+	} else {
+		roDbSettings := sa.DbSettings{
+			MaxOpenConns:    saConf.ReadOnlyDB.MaxOpenConns,
+			MaxIdleConns:    saConf.ReadOnlyDB.MaxIdleConns,
+			ConnMaxLifetime: saConf.ReadOnlyDB.ConnMaxLifetime.Duration,
+			ConnMaxIdleTime: saConf.ReadOnlyDB.ConnMaxIdleTime.Duration}
+
+		dbReadOnlyMap, err = sa.NewDbMap(dbReadOnlyURL, roDbSettings)
+		cmd.FailOnError(err, "Could not connect to read-only database")
+
+		dbReadOnlyAddr, dbReadOnlyUser, err := saConf.ReadOnlyDB.DSNAddressAndUser()
+		cmd.FailOnError(err, "Could not determine address or user of read-only DB DSN")
+
+		sa.InitDBMetrics(dbReadOnlyMap, scope, roDbSettings, dbReadOnlyAddr, dbReadOnlyUser)
```

Comment:
These lines are identical to lines 69-86 above. I'd factor them out into a tiny helper function (embedded as a closure here inside `main` so it is clear it has side effects) which takes a config section as input and returns a dbMap. Then the code would look approximately like
```
initDBMap := func(c cmd.DBConfig) *db.WrappedMap {
   ...
}

dbMap := initDBMap(saConf.DB)
dbReadOnlyMap := dbMap
if saConf.ReadOnlyDB.URL() == "" {
  dbReadOnlyMap = initDBMap(saConf.ReadOnlyDB)
}
```

Reply:
I'm making the same change to the ocsp-updater.